File: python-app/app/responses_loop.py
"""The complete Responses API → Python tool → tool results → response loop."""

from __future__ import annotations

import asyncio
import base64
import json
import os
import struct
import time
from typing import Any

from .errors import RunnerCoreError
from .processes import assert_active, await_active


class OpenAIResponsesClient:
    def __init__(self, api_key: str) -> None:
        from openai import AsyncOpenAI

        self.client = AsyncOpenAI(api_key=api_key)

    async def create(self, request: dict[str, Any], signal: asyncio.Event | None = None) -> dict[str, Any]:
        response = await await_active(self.client.responses.create(**request), signal)
        return response.model_dump(exclude_none=True)

    async def close(self) -> None:
        await self.client.close()


def create_default_responses_client() -> OpenAIResponsesClient:
    key = os.environ.get("OPENAI_API_KEY")
    if not key:
        raise RunnerCoreError(
            "OPENAI_API_KEY is required to run a lab.",
            code="missing_api_key",
            status_code=400,
            hint="Set OPENAI_API_KEY in the runner environment, then rerun the scenario.",
        )
    return OpenAIResponsesClient(key)


def build_code_tool_definitions(platform: str) -> list[dict[str, Any]]:
    return [
        {
            "type": "function",
            "name": "exec_py",
            "strict": True,
            "description": "Execute Python in the persistent PyAutoGUI desktop session for this run.",
            "parameters": {
                "type": "object",
                "additionalProperties": False,
                "required": ["code"],
                "properties": {
                    "code": {
                        "type": "string",
                        "description": "\n".join(
                            [
                                f"Python to execute on the local {platform} desktop. Python globals survive between calls.",
                                "Available globals: pyautogui, log(value), and display(image), accepting a Pillow image.",
                                "A visible Chromium window is already open to the lab. Start with display(pyautogui.screenshot()).",
                                "Use PyAutoGUI mouse and keyboard input. On macOS (darwin), use command hotkeys; elsewhere use ctrl.",
                                "Screenshots use the same coordinates as PyAutoGUI input, including on Retina displays. All coordinates refer to the full desktop screenshot.",
                                "Operate only the lab window. No Playwright objects are available. Do not change the PyAutoGUI fail-safe setting.",
                            ]
                        ),
                    }
                },
            },
        }
    ]


def invalid_response(message: str) -> RunnerCoreError:
    return RunnerCoreError(message, code="unexpected_model_response", status_code=400)


def classify_response(response: dict[str, Any]) -> dict[str, Any]:
    """Validate *all* output items before allowing any model code to execute."""
    if not isinstance(response, dict):
        raise invalid_response("Malformed Responses API response.")
    if response.get("status") != "completed" or response.get("error"):
        error = response.get("error")
        message = error.get("message", "") if isinstance(error, dict) else ""
        raise invalid_response(
            f"Responses API request did not complete (status: {response.get('status', 'missing')}). {message}".strip()
        )
    if not isinstance(response.get("id"), str) or not response["id"].strip():
        raise invalid_response("Responses API response ID is missing.")
    if not isinstance(response.get("output"), list):
        raise invalid_response("Responses API output is missing.")
    calls: list[dict[str, Any]] = []
    call_ids: set[str] = set()
    commentary: list[str] = []
    final: list[str] = []
    refusal: str | None = None
    for item in response["output"]:
        if not isinstance(item, dict):
            raise invalid_response("Malformed response output item.")
        if item.get("type") == "reasoning":
            continue
        if item.get("type") == "function_call":
            if item.get("name") != "exec_py":
                raise invalid_response(
                    f"Unexpected function call: {item.get('name', '<unknown>')}. Only exec_py is supported."
                )
            if "status" in item and item["status"] != "completed":
                raise invalid_response("The response contains an unfinished function call.")
            if not isinstance(item.get("call_id"), str) or not item["call_id"].strip():
                raise invalid_response("Function call ID is missing.")
            if item["call_id"] in call_ids:
                raise invalid_response(f"Duplicate function call ID: {item['call_id']}.")
            if not isinstance(item.get("arguments"), str):
                raise invalid_response("Function call arguments must be a JSON string.")
            try:
                args = json.loads(item["arguments"])
            except ValueError as error:
                raise invalid_response("Function call arguments are not valid JSON.") from error
            if not isinstance(args, dict) or set(args) != {"code"} or not isinstance(args.get("code"), str):
                raise invalid_response("exec_py requires a code string and no other arguments.")
            if not args["code"].strip() or len(args["code"].encode()) > 64 * 1024:
                raise invalid_response("Python code must be nonempty and at most 64 KiB.")
            call_ids.add(item["call_id"])
            calls.append({"call_id": item["call_id"], "arguments": item["arguments"], "code": args["code"]})
            continue
        if item.get("type") != "message":
            raise RunnerCoreError(
                f"Unsupported response output: {item.get('type')}. Only exec_py tool calls are supported.",
                code="unsupported_tool_response",
                status_code=400,
            )
        if (
            item.get("role") != "assistant"
            or not isinstance(item.get("content"), list)
            or ("status" in item and item["status"] != "completed")
        ):
            raise invalid_response("The response contains an invalid or unfinished assistant message.")
        if item.get("phase") not in (None, "commentary", "final_answer"):
            raise invalid_response("Unknown assistant message phase.")
        for part in item["content"]:
            if not isinstance(part, dict):
                raise invalid_response("Malformed assistant message content.")
            if part.get("type") == "refusal":
                text = part.get("refusal")
                refusal = text.strip() if isinstance(text, str) and text.strip() else "The model declined this task."
                continue
            if part.get("type") != "output_text" or not isinstance(part.get("text"), str):
                raise invalid_response("Unsupported assistant message content.")
            text = part["text"].strip()
            if text:
                (commentary if item.get("phase") == "commentary" else final).append(text)
    if refusal:
        raise RunnerCoreError(refusal, code="model_refusal", status_code=400)
    progress = "\n\n".join(commentary)
    if calls:
        return {"kind": "calls", "calls": calls, "commentary": progress}
    if final:
        return {"kind": "final", "text": "\n\n".join(final), "commentary": progress}
    if commentary:
        return {"kind": "commentary", "commentary": progress}
    raise invalid_response("Responses API returned no tool calls or nonempty final assistant message.")


async def execute_python_tool_call(context: Any, session: Any, call: dict[str, Any]) -> list[dict[str, Any]]:
    if session.execution is None:
        raise RuntimeError("Python runtime is unavailable. Run uv sync and start a new run.")
    await await_active(
        context.emit_event(
            {
                "detail": f"exec_py {call['arguments']}",
                "level": "pending",
                "message": "Function tool call received from the model.",
                "type": "function_call_requested",
            }
        ),
        context.signal,
    )
    result = await await_active(session.execution.execute(call["code"], context.signal), context.signal)
    assert_active(context.signal)
    for item in result["output"]:
        if item["type"] != "input_image":
            continue
        encoded = item["image_url"][len("data:image/png;base64,") :]
        image = base64.b64decode(encoded)
        observation: dict[str, Any] = {"base64": encoded}
        if len(image) >= 24 and image[:8] == b"\x89PNG\r\n\x1a\n":
            observation["width"], observation["height"] = struct.unpack(">II", image[16:24])
        await await_active(context.capture_screenshot(session, "Code tool image", observation), context.signal)
    await await_active(context.sync_browser_state(session), context.signal)
    await await_active(
        context.capture_screenshot(session, f"responses-code-turn-{time.time_ns() // 1_000_000}"), context.signal
    )
    await await_active(
        context.emit_event(
            {
                "detail": "exec_py",
                "level": "ok",
                "message": "Function tool call completed.",
                "type": "function_call_completed",
            }
        ),
        context.signal,
    )
    return result["output"] or [{"text": "exec_py completed with no output.", "type": "input_text"}]


async def run_responses_code_loop(
    context: Any,
    session: Any,
    instructions: str,
    max_response_turns: int,
    client: Any,
) -> dict[str, Any]:
    """Each iteration requests a response and sends all matching tool results back."""
    previous_response_id: str | None = None
    next_input: Any = context.detail["run"]["prompt"].strip()
    final_assistant_message: str | None = None
    for turn in range(1, max_response_turns + 1):
        assert_active(context.signal)
        request: dict[str, Any] = {
            "instructions": instructions,
            "input": next_input,
            "model": context.detail["run"]["model"],
            "parallel_tool_calls": False,
            "reasoning": {"effort": "low"},
            "truncation": "auto",
            "tools": build_code_tool_definitions(session.execution.platform),
        }
        if previous_response_id is not None:
            request["previous_response_id"] = previous_response_id
        response = await await_active(client.create(request, context.signal), context.signal)
        assert_active(context.signal)
        classified = classify_response(response)
        usage = response.get("usage") or {}
        reasoning = (usage.get("output_tokens_details") or {}).get("reasoning_tokens", 0)
        await await_active(
            context.emit_event(
                {
                    "detail": f"{response['id']} · {usage.get('input_tokens', 0)} in · {usage.get('output_tokens', 0)} out · {reasoning} reasoning",
                    "level": "ok",
                    "message": f"Responses API turn {turn} completed.",
                    "type": "run_progress",
                }
            ),
            context.signal,
        )
        previous_response_id = response["id"]
        if classified["commentary"]:
            await await_active(
                context.emit_event(
                    {
                        "type": "run_progress",
                        "level": "pending",
                        "message": "Model progress.",
                        "detail": classified["commentary"],
                    }
                ),
                context.signal,
            )
        if classified["kind"] == "final":
            final_assistant_message = classified["text"]
            break
        if classified["kind"] == "commentary":
            next_input = []
            continue
        tool_outputs = []
        for call in classified["calls"]:
            assert_active(context.signal)
            output = await execute_python_tool_call(context, session, call)
            assert_active(context.signal)
            tool_outputs.append({"call_id": call["call_id"], "output": output, "type": "function_call_output"})
        next_input = tool_outputs
    if final_assistant_message is None:
        raise RuntimeError(
            f"Responses API code loop exhausted the configured {max_response_turns}-turn budget without producing a final assistant message."
        )
    await await_active(
        context.emit_event(
            {
                "detail": final_assistant_message,
                "level": "ok",
                "message": "Model returned a final response.",
                "type": "run_progress",
            }
        ),
        context.signal,
    )
    return {
        "final_assistant_message": final_assistant_message,
        "notes": [
            "Executed the scenario through a live Responses API code loop.",
            f"Model final response: {final_assistant_message}",
        ],
    }

File: python-app/packages/runner-core/src/responses-loop.ts
import OpenAI from "openai";

import { type BrowserSession } from "@cua-sample/browser-runtime";

import { RunnerCoreError } from "./errors.js";
import type { RunExecutionContext } from "./scenario-runtime.js";

type FunctionCallItem = {
  arguments?: string;
  call_id?: string;
  name?: string;
  status?: string;
  type: "function_call";
};

type PythonCall = {
  call_id: string;
  arguments: string;
  code: string;
};

type MessageItem = {
  content?: Array<{
    text?: string;
    refusal?: string;
    type?: string;
  }>;
  role?: string;
  status?: string;
  phase?: string | null;
  type: "message";
};

type ResponseOutputItem =
  | FunctionCallItem
  | MessageItem
  | { [key: string]: unknown; type: string };

type ResponsesApiResponse = {
  error?: { message?: string } | null;
  id: string;
  output?: ResponseOutputItem[];
  status?: string;
  usage?: {
    input_tokens?: number;
    output_tokens?: number;
    output_tokens_details?: {
      reasoning_tokens?: number;
    };
    total_tokens?: number;
  } | null;
};

type ResponsesLoopMode = "auto" | "fallback" | "live";

type ResponsesClient = {
  create: (
    request: Record<string, unknown>,
    signal: AbortSignal,
  ) => Promise<ResponsesApiResponse>;
};

type ResponsesLoopContext = {
  context: RunExecutionContext;
  instructions: string;
  maxResponseTurns: number;
  prompt?: string;
  session: BrowserSession;
};

type ResponsesLoopResult = {
  finalAssistantMessage?: string;
  notes: string[];
};

class OpenAIResponsesClient implements ResponsesClient {
  private readonly client: OpenAI;

  constructor(apiKey: string) {
    this.client = new OpenAI({ apiKey });
  }

  async create(request: Record<string, unknown>, signal: AbortSignal) {
    return (await this.client.responses.create(request, {
      signal,
    })) as ResponsesApiResponse;
  }
}

function assertActive(signal: AbortSignal) {
  if (signal.aborted) {
    throw new Error("Run aborted.");
  }
}

function parseResponsesLoopMode(env: NodeJS.ProcessEnv = process.env): ResponsesLoopMode {
  const raw = env.CUA_RESPONSES_MODE?.trim().toLowerCase();

  if (raw === "live" || raw === "fallback" || raw === "auto") {
    return raw;
  }

  return "auto";
}

function isTestEnvironment(env: NodeJS.ProcessEnv = process.env) {
  return env.NODE_ENV === "test" || env.VITEST === "true";
}

export function createDefaultResponsesClient(): ResponsesClient | null {
  const mode = parseResponsesLoopMode();
  const apiKey = process.env.OPENAI_API_KEY;

  if (mode === "fallback") {
    return null;
  }

  if (!apiKey) {
    if (mode === "live") {
      throw new RunnerCoreError(
        "CUA_RESPONSES_MODE=live requires OPENAI_API_KEY to be set.",
        {
          code: "missing_api_key",
          hint:
            "Set OPENAI_API_KEY before starting a live CUA run, or switch CUA_RESPONSES_MODE back to auto.",
          statusCode: 400,
        },
      );
    }

    return null;
  }

  if (mode === "auto" && isTestEnvironment()) {
    return null;
  }

  return new OpenAIResponsesClient(apiKey);
}

function describeUsage(response: ResponsesApiResponse) {
  const inputTokens = response.usage?.input_tokens ?? 0;
  const outputTokens = response.usage?.output_tokens ?? 0;
  const reasoningTokens = response.usage?.output_tokens_details?.reasoning_tokens ?? 0;

  return `${inputTokens} in · ${outputTokens} out · ${reasoningTokens} reasoning`;
}

async function emitModelTurnEvent(
  context: RunExecutionContext,
  response: ResponsesApiResponse,
  turn: number,
) {
  await context.emitEvent({
    detail: `${response.id} · ${describeUsage(response)}`,
    level: "ok",
    message: `Responses API turn ${turn} completed.`,
    type: "run_progress",
  });
}

function buildCodeToolDefinitions(platform: string) {
  return [
    {
      type: "function",
      name: "exec_py",
      description:
        "Execute Python in the persistent PyAutoGUI desktop session for this run.",
      strict: true,
      parameters: {
        additionalProperties: false,
        properties: {
          code: {
            description: [
              `Python to execute on the local ${platform} desktop. Python globals survive between calls.`,
              "Available globals: pyautogui, log(value), and display(image), accepting a Pillow image.",
              "A visible Chromium window is already open to the lab. Start with display(pyautogui.screenshot()).",
              "Use PyAutoGUI mouse and keyboard input. On macOS (darwin), use command hotkeys; elsewhere use ctrl.",
              "Screenshots use the same coordinates as PyAutoGUI input, including on Retina displays. All coordinates refer to the full desktop screenshot.",
              "Operate only the lab window. No Playwright objects are available. Do not change the PyAutoGUI fail-safe setting.",
            ].join("\n"),
            type: "string",
          },
        },
        required: ["code"],
        type: "object",
      },
    },
  ];
}

async function executePythonToolCall(
  input: ResponsesLoopContext,
  functionCall: PythonCall,
) {
  if (!input.session.execution) {
    throw new Error(
      "Python runtime is unavailable. Install runtimes/requirements.txt into .venv and start a new run.",
    );
  }
  const result = await input.session.execution.execute(functionCall.code, input.context.signal);
  for (const item of result.output) {
    if (item.type !== "input_image") continue;
    const base64 = item.image_url.slice("data:image/png;base64,".length);
    const bytes = Buffer.from(base64, "base64");
    const isPng = bytes.length >= 24 &&
      bytes.subarray(0, 8).equals(Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]));
    const dimensions = isPng
      ? { width: bytes.readUInt32BE(16), height: bytes.readUInt32BE(20) }
      : {};
    await input.context.captureScreenshot(input.session, "Code tool image", {
      base64, ...dimensions, source: "code_tool",
    });
  }
  await input.context.syncBrowserState(input.session);
  await input.context.captureScreenshot(
    input.session,
    `responses-code-turn-${Date.now()}`,
  );
  return result.output.length > 0
    ? result.output
    : [{ text: "exec_py completed with no output.", type: "input_text" as const }];
}

async function executeFunctionToolCall(
  input: ResponsesLoopContext,
  functionCall: PythonCall,
) {
  await input.context.emitEvent({
    detail: `exec_py ${functionCall.arguments}`,
    level: "pending",
    message: "Function tool call received from the model.",
    type: "function_call_requested",
  });

  const output = await executePythonToolCall(input, functionCall);

  await input.context.emitEvent({
    detail: "exec_py",
    level: "ok",
    message: "Function tool call completed.",
    type: "function_call_completed",
  });

  return output;
}

function invalidResponse(message: string) {
  return new RunnerCoreError(message, {
    code: "unexpected_model_response",
    statusCode: 400,
  });
}

// Validate every output item before allowing any Python to execute.
function classifyResponse(response: ResponsesApiResponse) {
  if (response.status !== "completed" || response.error) {
    throw invalidResponse(
      `Responses API request did not complete (status: ${response.status ?? "missing"}). ${response.error?.message ?? ""}`.trim(),
    );
  }
  if (!Array.isArray(response.output)) {
    throw invalidResponse("Responses API output is missing.");
  }

  const calls: PythonCall[] = [];
  const callIds = new Set<string>();
  const commentary: string[] = [];
  const final: string[] = [];
  let refusal: string | undefined;

  for (const item of response.output) {
    if (!item || typeof item !== "object") {
      throw invalidResponse("Malformed response output item.");
    }
    if (item.type === "reasoning") continue;

    if (item.type === "function_call") {
      const call = item as FunctionCallItem;
      if (call.name !== "exec_py") {
        throw invalidResponse(`Unexpected function call: ${call.name ?? "<unknown>"}. Only exec_py is supported.`);
      }
      if (call.status !== undefined && call.status !== "completed") {
        throw invalidResponse("The response contains an unfinished function call.");
      }
      if (typeof call.call_id !== "string" || !call.call_id.trim()) {
        throw invalidResponse("Function call ID is missing.");
      }
      if (callIds.has(call.call_id)) {
        throw invalidResponse(`Duplicate function call ID: ${call.call_id}.`);
      }
      if (typeof call.arguments !== "string") {
        throw invalidResponse("Function call arguments must be a JSON string.");
      }
      let args: unknown;
      try {
        args = JSON.parse(call.arguments);
      } catch {
        throw invalidResponse("Function call arguments are not valid JSON.");
      }
      if (
        !args || typeof args !== "object" || Array.isArray(args) ||
        !("code" in args) || typeof args.code !== "string" ||
        Object.keys(args).some((key) => key !== "code")
      ) {
        throw invalidResponse("exec_py requires a code string and no other arguments.");
      }
      if (!args.code.trim() || Buffer.byteLength(args.code) > 64 * 1024) {
        throw invalidResponse("Python code must be nonempty and at most 64 KiB.");
      }
      callIds.add(call.call_id);
      calls.push({ call_id: call.call_id, arguments: call.arguments, code: args.code });
      continue;
    }

    if (item.type !== "message") {
      throw new RunnerCoreError(`Unsupported response output: ${item.type}. Only exec_py tool calls are supported.`, {
        code: "unsupported_tool_response",
        statusCode: 400,
      });
    }
    const message = item as MessageItem;
    if (
      message.role !== "assistant" || !Array.isArray(message.content) ||
      (message.status !== undefined && message.status !== "completed")
    ) {
      throw invalidResponse("The response contains an invalid or unfinished assistant message.");
    }
    if (message.phase != null && message.phase !== "commentary" && message.phase !== "final_answer") {
      throw invalidResponse("Unknown assistant message phase.");
    }
    for (const part of message.content) {
      if (!part || typeof part !== "object") {
        throw invalidResponse("Malformed assistant message content.");
      }
      if (part.type === "refusal") {
        refusal = (typeof part.refusal === "string" && part.refusal.trim()) || "The model declined this task.";
        continue;
      }
      if (part.type !== "output_text" || typeof part.text !== "string") {
        throw invalidResponse("Unsupported assistant message content.");
      }
      if (!part.text.trim()) continue;
      if (message.phase === "commentary") {
        commentary.push(part.text.trim());
      } else {
        // Phase is optional per message, including alongside explicit commentary.
        final.push(part.text.trim());
      }
    }
  }

  if (refusal) {
    throw new RunnerCoreError(refusal, { code: "model_refusal", statusCode: 400 });
  }
  const progress = commentary.join("\n\n");
  if (calls.length) return { kind: "calls" as const, calls, commentary: progress };
  if (final.length) return { kind: "final" as const, text: final.join("\n\n"), commentary: progress };
  if (commentary.length) return { kind: "commentary" as const, commentary: progress };
  throw invalidResponse("Responses API returned no tool calls or nonempty final assistant message.");
}

export async function runResponsesCodeLoop(
  input: ResponsesLoopContext,
  client: ResponsesClient,
): Promise<ResponsesLoopResult> {
  let previousResponseId: string | undefined;
  let nextInput: unknown = input.prompt ?? input.context.detail.run.prompt;
  let finalAssistantMessage: string | undefined;

  for (let turn = 1; turn <= input.maxResponseTurns; turn += 1) {
    assertActive(input.context.signal);
    const response = await client.create(
      {
        instructions: input.instructions,
        input: nextInput,
        model: input.context.detail.run.model,
        parallel_tool_calls: false,
        previous_response_id: previousResponseId,
        reasoning: { effort: "low" },
        tools: buildCodeToolDefinitions(input.session.execution?.platform ?? process.platform),
        truncation: "auto",
      },
      input.context.signal,
    );
    assertActive(input.context.signal);
    const classified = classifyResponse(response);
    await emitModelTurnEvent(input.context, response, turn);

    previousResponseId = response.id;
    if (classified.commentary) {
      await input.context.emitEvent({
        type: "run_progress",
        level: "pending",
        message: "Model progress.",
        detail: classified.commentary,
      });
    }
    if (classified.kind === "final") {
      finalAssistantMessage = classified.text;
      break;
    }
    if (classified.kind === "commentary") {
      nextInput = [];
      continue;
    }
    const functionCalls = classified.calls;

    const toolOutputs = [];

    for (const functionCall of functionCalls) {
      assertActive(input.context.signal);
      const output = await executeFunctionToolCall(input, functionCall);

      toolOutputs.push({
        call_id: functionCall.call_id,
        output,
        type: "function_call_output",
      });
    }

    nextInput = toolOutputs;
  }

  if (!finalAssistantMessage) {
    throw new Error(
      `Responses API code loop exhausted the configured ${input.maxResponseTurns}-turn budget without producing a final assistant message.`,
    );
  }

  await input.context.emitEvent({
    detail: finalAssistantMessage,
    level: "ok",
    message: "Model returned a final response.",
    type: "run_progress",
  });

  return {
    finalAssistantMessage,
    notes: [
      "Executed the scenario through a live Responses API code loop.",
      `Model final response: ${finalAssistantMessage}`,
    ],
  };
}
